### ISSUE TITLE
Adding Text to Speech feature using Mozilla Deepspeech

### ISSUE THREAD

**Issue by fux0r2009** [florisboard/florisboard#195]
It would be awesome if you could add support for text to speech to Florisboard using **Mozilla DeepSpeech** Technology readily available for android : [](https://deepspeech.readthedocs.io/en/latest/SUPPORTED_PLATFORMS.html#android-aarch64).

It could be part of the core or preferably a module to the add at the discretion of the user. So let me know what you think...

And by the way, excellent work as this is the best Open source keyboard I ever used :) keep it up champion!

**Comment by patrickgold** [comment]
Thanks for your suggestion! Mozilla's DeepSpeech program seems pretty cool and I will definitely play around with it! I will probably include it in FlorisBoard when I have the time to do so. The inclusion will most likely happen as an extension pack (to keep the core APK size small), so I cannot include DeepSpech before a proper extension support is existent in FlorisBoard. I will place it on the todo list for v0.5.0!

**Comment by fux0r2009** [comment]
Thank you very much, I will look forward to it's implementation and again Excellent work on making **florisboard** the best open source keyboard :) keep up the good work!

**Comment by parcelcat** [comment]
This is a great idea, and I think speech-to-text would be best implemented as a voice input provider for Android. This way, other apps could also use the functionality. Whether the voice input provider should be bundled with FlorisBoard or split into another app might need some discussion.

**Comment by joe-narvaez** [comment]
Is there anyway i can sponsor this feature?

**Comment by iot-resister** [comment]
Is there a way i can sponsor this feature?

**Comment by Tombstone2K** [comment]
It seems that VOSK is much better at understanding a wide variety of accents and has a higher success rate.
It'd much better to integrate VOSK to florisboard.

@Felicis made this app https://github.com/Felicis/vosk-android-demo which acts as a voice-only typing keyboard.
Maybe that can be integrated into florisboard, i.e. (a microphone button next to the paste button which opens up this linked apps UI) 
PS: The linked app is missing some features, like a space button.

What do y'all think ??

**Comment by parcelcat** [comment]
> Maybe in the future, florisboard can give an option to choose a RecognitionService, so it'd be compatible with whichever service the user wants...

Yes, this is a great idea. It's a lot more work to create a speech-to-text solution than to interface with an existing one that is already installed on the device.

**Comment by ElishaAz** [comment]
I've created a fork with a button that switches to a keyboard that supports voice input (this is what Gboard/ AOSP keyboard does): [https://github.com/ElishaAz/florisboard/tree/switch-to-voice-ime](https://github.com/ElishaAz/florisboard/tree/switch-to-voice-ime).

You'll need to install a keyboard that has the voice input mode. [I built one using Vosk](https://github.com/ElishaAz/Vosk-Voice-IME/releases/tag/v1.0.0). Other options are [Google Speech Services](https://play.google.com/store/apps/details?id=com.google.android.tts), and [Konele](https://github.com/Kaljurand/K6nele/).

I started working on integrating voice input inside the keyboard using [SpeechRecognizer](https://developer.android.com/reference/android/speech/SpeechRecognizer).

**Comment by patrickgold** [comment]
@ElishaAz I just had a look of your changes in your fork and it looks really clean and well done to me! This would be the best solution for now, because almost every device has at least one IME with voice input support installed.

Could you please make a PR with the changes, so I can properly review and test them for merging into this project's code base?

**Comment by ElishaAz** [comment]
Created a PR: https://github.com/florisboard/florisboard/pull/1587

**Comment by Lastofthefirst** [comment]
@ElishaAz Your Vosk keyboard (https://github.com/ElishaAz/Vosk-Voice-IME/releases/tag/v1.0.0) is exactly what I am looking for! What would it take to polish the ui a little further? Im not familiar with java. I feel that with a few small adjustments, or i hope they would be small, the setup would be very viable for a wide audience.

Maybe,

* changing the blue to a more muted grey or black, maybe match florisboard?
* adding icons if its not too difficult, maybe even emoji if icons are too much work, just to reduce the cognitive load; Mic/recording/stop recording/loading 🎙 🎤 🤫 🤐  and while its still loading  🔇 🔕 🛑 🙉.
* Could also give it a unique name 

Anyway I'm so happy with this, thank you, I think its worth drawing attention to it, I searched around a good deal before finding it.

**Comment by ElishaAz** [comment]
Honestly, I copied the UI from a different project, and hadn't worked on it at all. I'll work on it in the next few days.
As for the name, I'm open for suggestions.

If you have any other suggestions / improvements, please open an issue in my project.

**Comment by ElishaAz** [comment]
> @ElishaAz Your Vosk keyboard (https://github.com/ElishaAz/Vosk-Voice-IME/releases/tag/v1.0.0) is exactly what I am looking for! What would it take to polish the ui a little further? Im not familiar with java. I feel that with a few small adjustments, or i hope they would be small, the setup would be very viable for a wide audience.
> 
> Maybe,
> 
>     * changing the blue to a more muted grey or black, maybe match florisboard?
> 
>     * adding icons if its not too difficult, maybe even emoji if icons are too much work, just to reduce the cognitive load; Mic/recording/stop recording/loading studio_microphone microphone shushing_face zipper_mouth_face  and while its still loading  mute no_bell stop_sign hear_no_evil.
> 
>     * Could also give it a unique name
> 
> 
> Anyway I'm so happy with this, thank you, I think its worth drawing attention to it, I searched around a good deal before finding it.

Here it is: [Sayboard](https://github.com/ElishaAz/Sayboard/releases/tag/v2.0.0)

**Comment by S7venLights** [comment]
More projects like this: 

https://github.com/ElishaAz/Sayboard/issues/6

**Comment by bertin0** [comment]
Check out this project too: https://github.com/abb128/WhisperVoiceInput . It's made by a guy who works on live captions for Linux.

**Comment by grigio** [comment]
Can you add the support to vosk android service tts ? it works very well with anysoftkeyboard
https://github.com/alphacep/vosk-android-service

**Comment by S7venLights** [comment]
The VOSK STT you mean. 
They are already working on that and… 
It already works with androids built in IME Method I believe if you install any of these 👇 and then click the mic button in keyboards: 

For TTS, get RHVoice. Works with Maps

for local STT use Sayboard:
https://github.com/elishaaz/vosk-voice-ime (Works from your keyboard when clicking the mic but has no punctuation yet)

Or… 

German and English IME (works with keyboard mic button) with useful punctuation buttons. ￼
Abandoned but good: 
https://github.com/Felicis/vosk-android-demo

Server based STT: 
I just discovered this, haven't tested yet (Not sure how private it is): 
https://github.com/Kaljurand/K6nele

**Comment by soupslurpr** [comment]
I think it'd be great if FlorisBoard could use the SpeechRecognizer class for this feature. It would allow it to use other apps as well as it's own implementation. The user can set an app as their preferred implementation which is what SpeechRecognizer uses by default, but it can be overridden to use specific apps or even an internal implementation.

The initial implementation could just use the user's preferred app for it which wouldn't increase APK size a lot.

I think this is the most flexible way to implement this feature.

For UI I'm not sure, but maybe a minimal approach like Gboard would be best where the voice recognition is just a small mic button at the top right of the keyboard, and the rest of the keyboard is still visible. This would allow users to use the keyboard and its features at the same time.

Please let me know when implementing this starts, I'm interested in helping (if needed) if FlorisBoard takes this route (even if it's just in the implementation and not the UI).
